File: packages/nativescript-image-cache-it/src-native/imagecacheit/src/main/java/com/github/triniwiz/imagecacheit/MyAppGlideModule.java
package com.github.triniwiz.imagecacheit;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.bumptech.glide.GlideBuilder;
import com.bumptech.glide.Registry;
import com.bumptech.glide.annotation.Excludes;
import com.bumptech.glide.annotation.GlideModule;
import com.bumptech.glide.integration.okhttp3.*;
import com.bumptech.glide.load.engine.bitmap_recycle.LruBitmapPool;
import com.bumptech.glide.load.engine.cache.InternalCacheDiskCacheFactory;
import com.bumptech.glide.load.engine.cache.LruResourceCache;
import com.bumptech.glide.load.engine.cache.MemorySizeCalculator;
import com.bumptech.glide.load.model.GlideUrl;
import com.bumptech.glide.module.AppGlideModule;

import okhttp3.*;
import okio.Buffer;
import okio.BufferedSource;
import okio.ForwardingSource;
import okio.Okio;
import okio.Source;

import com.bumptech.glide.load.ImageHeaderParser;
import com.bumptech.glide.load.resource.bitmap.ExifInterfaceImageHeaderParser;

import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;
import java.util.WeakHashMap;

@GlideModule
@Excludes(OkHttpLibraryGlideModule.class)
public final class MyAppGlideModule extends AppGlideModule {

    private static DispatchingProgressListener progressListener = new DispatchingProgressListener();

    @Override
    public void registerComponents(@NonNull Context context, @NonNull Glide glide, @NonNull Registry registry) {
        super.registerComponents(context, glide, registry);
        OkHttpClient okHttpClient = new OkHttpClient();
        registry.replace(GlideUrl.class, InputStream.class, new OkHttpUrlLoader.Factory(okHttpClient));



        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(createInterceptor(progressListener))
                .build();
        OkHttpUrlLoader.Factory factory = new OkHttpUrlLoader.Factory(client);
        registry.replace(GlideUrl.class, InputStream.class, factory);

        for (ImageHeaderParser parser : registry.getImageHeaderParsers()) {
            if (parser instanceof ExifInterfaceImageHeaderParser) {
                registry.getImageHeaderParsers().remove(parser);
            }
        }
    }

    @Override
    public void applyOptions(@NonNull Context context, @NonNull GlideBuilder builder) {
        int diskCacheSizeBytes = 1024 * 1024 * 1000; // 1GB
        builder.setDiskCache(new InternalCacheDiskCacheFactory(context, diskCacheSizeBytes));
        /*MemorySizeCalculator calculator = new MemorySizeCalculator.Builder(context)
                .setMemoryCacheScreens(4)
                .build();
        builder.setMemoryCache(new LruResourceCache(calculator.getMemoryCacheSize()));
        MemorySizeCalculator bpCalculator = new MemorySizeCalculator.Builder(context)
                .setBitmapPoolScreens(3)
                .build();
        builder.setBitmapPool(new LruBitmapPool(bpCalculator.getBitmapPoolSize()));
        */
    }



    private static Interceptor createInterceptor(final ResponseProgressListener listener) {
        return new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();
                Response response = chain.proceed(request);
                final String key = request.url().toString();
                return response
                        .newBuilder()
                        .body(new OkHttpProgressResponseBody(key, response.body(), listener))
                        .build();
            }
        };
    }

    static void forget(String key) {
        progressListener.forget(key);
    }

    static void expect(String key, ImageViewProgressListener listener) {
        progressListener.expect(key, listener);
    }

    private interface ResponseProgressListener {
        void update(String key, long bytesRead, long contentLength);
    }

    private static class DispatchingProgressListener implements ResponseProgressListener {
        private final Map<String, ImageViewProgressListener> LISTENERS = new WeakHashMap<>();
        private final Map<String, Long> PROGRESSES = new HashMap<>();

        private final Handler handler;

        DispatchingProgressListener() {
            this.handler = new Handler(Looper.getMainLooper());
        }

        void forget(String key) {
            LISTENERS.remove(key);
            PROGRESSES.remove(key);
        }

        void expect(String key, ImageViewProgressListener listener) {
            LISTENERS.put(key, listener);
        }

        @Override
        public void update(final String key, final long bytesRead, final long contentLength) {
            final ImageViewProgressListener listener = LISTENERS.get(key);
            if (listener == null) {
                return;
            }
            if (contentLength <= bytesRead) {
                forget(key);
            }
            if (needsDispatch(key, bytesRead, contentLength, listener.getGranularityPercentage())) {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        listener.onProgress(key, bytesRead, contentLength);
                    }
                });
            }
        }

        private boolean needsDispatch(String key, long current, long total, float granularity) {
            if (granularity == 0 || current == 0 || total == current) {
                return true;
            }
            float percent = 100f * current / total;
            long currentProgress = (long) (percent / granularity);
            Long lastProgress = PROGRESSES.get(key);
            if (lastProgress == null || currentProgress != lastProgress) {
                PROGRESSES.put(key, currentProgress);
                return true;
            } else {
                return false;
            }
        }
    }

    private static class OkHttpProgressResponseBody extends ResponseBody {
        private final String key;
        private final ResponseBody responseBody;
        private final ResponseProgressListener progressListener;
        private BufferedSource bufferedSource;

        OkHttpProgressResponseBody(
                String key,
                ResponseBody responseBody,
                ResponseProgressListener progressListener
        ) {
            this.key = key;
            this.responseBody = responseBody;
            this.progressListener = progressListener;
        }

        @Override
        public MediaType contentType() {
            return responseBody.contentType();
        }

        @Override
        public long contentLength() {
            return responseBody.contentLength();
        }

        @Override
        public BufferedSource source() {
            if (bufferedSource == null) {
                bufferedSource = Okio.buffer(source(responseBody.source()));
            }
            return bufferedSource;
        }

        private Source source(Source source) {
            return new ForwardingSource(source) {
                long totalBytesRead = 0L;

                @Override
                public long read(Buffer sink, long byteCount) throws IOException {
                    long bytesRead = super.read(sink, byteCount);
                    long fullLength = responseBody.contentLength();
                    if (bytesRead == -1) {
                        // this source is exhausted
                        totalBytesRead = fullLength;
                    } else {
                        totalBytesRead += bytesRead;
                    }
                    progressListener.update(key, totalBytesRead, fullLength);
                    return bytesRead;
                }
            };
        }
    }
}

File: packages/nativescript-image-cache-it/src-native/imagecacheit/src/main/java/com/github/triniwiz/imagecacheit/MyAppGlideModule.kt
package com.github.triniwiz.imagecacheit

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.bumptech.glide.Glide
import com.bumptech.glide.GlideBuilder
import com.bumptech.glide.Registry
import com.bumptech.glide.annotation.Excludes
import com.bumptech.glide.annotation.GlideModule
import com.bumptech.glide.integration.okhttp3.OkHttpLibraryGlideModule
import com.bumptech.glide.integration.okhttp3.OkHttpUrlLoader
import com.bumptech.glide.load.engine.bitmap_recycle.LruBitmapPool
import com.bumptech.glide.load.engine.cache.InternalCacheDiskCacheFactory
import com.bumptech.glide.load.engine.cache.LruResourceCache
import com.bumptech.glide.load.model.GlideUrl
import com.bumptech.glide.load.resource.bitmap.ExifInterfaceImageHeaderParser
import com.bumptech.glide.module.AppGlideModule
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import okio.*
import java.io.IOException
import java.io.InputStream
import java.util.*

@GlideModule
@Excludes(OkHttpLibraryGlideModule::class)
class MyAppGlideModule : AppGlideModule() {
  override fun registerComponents(context: Context, glide: Glide, registry: Registry) {
    super.registerComponents(context, glide, registry)
    val okHttpClient = OkHttpClient()
    registry.replace(GlideUrl::class.java, InputStream::class.java, OkHttpUrlLoader.Factory(okHttpClient))
    val client = OkHttpClient.Builder()
      .addInterceptor(createInterceptor(progressListener))
      .build()
    val factory = OkHttpUrlLoader.Factory(client)
    registry.replace(GlideUrl::class.java, InputStream::class.java, factory)
    for (parser in registry.imageHeaderParsers) {
      if (parser is ExifInterfaceImageHeaderParser) {
        registry.imageHeaderParsers.remove(parser)
      }
    }
  }

  override fun applyOptions(context: Context, builder: GlideBuilder) {
    if (maxDiskCacheSize > 0) {
      builder.setDiskCache(InternalCacheDiskCacheFactory(context, maxDiskCacheSize))
    }
    if (maxMemoryCacheSize > 0) {
      builder.setMemoryCache(LruResourceCache(maxMemoryCacheSize))
      builder.setBitmapPool(LruBitmapPool(maxMemoryCacheSize))
    }
  }

  private interface ResponseProgressListener {
    fun update(key: String, bytesRead: Long, contentLength: Long)
  }

  private class DispatchingProgressListener internal constructor() : ResponseProgressListener {
    private val LISTENERS: MutableMap<String, ImageViewProgressListener> = WeakHashMap()
    private val PROGRESSES: MutableMap<String, Long> = HashMap()
    private val handler: Handler = Handler(Looper.getMainLooper())
    fun forget(key: String) {
      LISTENERS.remove(key)
      PROGRESSES.remove(key)
    }

    fun expect(key: String, listener: ImageViewProgressListener) {
      LISTENERS[key] = listener
    }

    override fun update(key: String, bytesRead: Long, contentLength: Long) {
      val listener = LISTENERS[key] ?: return
      if (contentLength <= bytesRead) {
        forget(key)
      }
      if (needsDispatch(key, bytesRead, contentLength, listener.granularityPercentage)) {
        handler.post { listener.onProgress(key, bytesRead, contentLength) }
      }
    }

    private fun needsDispatch(key: String, current: Long, total: Long, granularity: Float): Boolean {
      if (granularity == 0f || current == 0L || total == current) {
        return true
      }
      val percent = 100f * current / total
      val currentProgress = (percent / granularity).toLong()
      val lastProgress = PROGRESSES[key]
      return if (lastProgress == null || currentProgress != lastProgress) {
        PROGRESSES[key] = currentProgress
        true
      } else {
        false
      }
    }

  }

  private class OkHttpProgressResponseBody internal constructor(
    private val key: String,
    private val responseBody: ResponseBody?,
    private val progressListener: ResponseProgressListener
  ) : ResponseBody() {
    private var bufferedSource: BufferedSource? = null
    override fun contentType(): MediaType? {
      return responseBody!!.contentType()
    }

    override fun contentLength(): Long {
      return responseBody!!.contentLength()
    }

    override fun source(): BufferedSource {
      if (bufferedSource == null) {
        bufferedSource = Okio.buffer(source(responseBody!!.source()))
      }
      return bufferedSource!!
    }

    private fun source(source: Source): Source {
      return object : ForwardingSource(source) {
        var totalBytesRead = 0L
        @Throws(IOException::class)
        override fun read(sink: Buffer, byteCount: Long): Long {
          val bytesRead = super.read(sink, byteCount)
          val fullLength = responseBody!!.contentLength()
          if (bytesRead == -1L) {
            // this source is exhausted
            totalBytesRead = fullLength
          } else {
            totalBytesRead += bytesRead
          }
          progressListener.update(key, totalBytesRead, fullLength)
          return bytesRead
        }
      }
    }
  }

  companion object {
    private val progressListener = DispatchingProgressListener()
    @JvmStatic
    var maxDiskCacheSize: Long = 0
    @JvmStatic
    var maxMemoryCacheSize: Long = 0
    @JvmStatic
    var maxDiskCacheAge: Long = 0
    private fun createInterceptor(listener: ResponseProgressListener): Interceptor {
      return Interceptor { chain ->
        val request = chain.request()
        val response = chain.proceed(request)
        val key = request.url().toString()
        response
          .newBuilder()
          .body(OkHttpProgressResponseBody(key, response.body(), listener))
          .build()
      }
    }

    fun forget(key: String) {
      progressListener.forget(key)
    }

    fun expect(key: String, listener: ImageViewProgressListener) {
      progressListener.expect(key, listener)
    }
  }
}
